qcodes/instrument_drivers/Spectrum/M4i.py

```diff
@@ -0,0 +1,691 @@
+
+        # data acquisition
+        self._def_transfer64bit(pyspcm.SPCM_BUF_DATA, pyspcm.SPCM_DIR_CARDTOPC, 0, data_pointer, 0, 2 * memsize)
+        self.general_command(pyspcm.M2CMD_DATA_STARTDMA | pyspcm.M2CMD_DATA_WAITDMA)
+
+        # convert buffer to numpy array
+        data = ct.cast(data_pointer, ct.POINTER(buffer_size))
+        output = np.frombuffer(data.contents, dtype=ct.c_int16)
+
+        self._stop_acquisition()
+
+        voltages = self.convert_to_voltage(output, mV_range / 1000)
+
+        return voltages
+
+    def close(self):
+        """Close handle to the card."""
+        if self.hCard is not None:
+            pyspcm.spcm_vClose(self.hCard)
+            self.hCard = None
+        super().close()
+
+    def get_card_type(self, verbose=0):
+        """Read card type."""
+        # read type, function and sn and check for D/A card
+        lCardType = pyspcm.int32(0)
+        pyspcm.spcm_dwGetParam_i32(self.hCard, pyspcm.SPC_PCITYP, pyspcm.byref(lCardType))
+        if verbose:
+            print('card_type: %s' % szTypeToName(lCardType.value))
+        return (lCardType.value)
+
+    # only works if the error was not caused by running the entire program (and therefore making a new M4i object)
+    def get_error_info32bit(self):
+        """Read an error from the error register."""
+        dwErrorReg = pyspcm.uint32(0)
+        lErrorValue = pyspcm.int32(0)
+
+        pyspcm.spcm_dwGetErrorInfo_i32(self.hCard, pyspcm.byref(dwErrorReg), pyspcm.byref(lErrorValue), None)
+        return (dwErrorReg.value, lErrorValue.value)
+
+    def _param64bit(self, param):
+        """Read a 64-bit parameter from the device."""
+        data = pyspcm.int64(0)
+        pyspcm.spcm_dwGetParam_i64(self.hCard, param, pyspcm.byref(data))
+        return (data.value)
+
+    def _param32bit(self, param):
+        """Read a 32-bit parameter from the device."""
+        data = pyspcm.int32(0)
+        pyspcm.spcm_dwGetParam_i32(self.hCard, param, pyspcm.byref(data))
+        return (data.value)
+
+    def _set_param32bit(self, param, value):
+        """Read a 32-bit parameter from the device."""
+        pyspcm.spcm_dwSetParam_i32(self.hCard, param, value)
+
+    def _invalidate_buf(self, buf_type):
+        """Invalidate device buffer."""
+        pyspcm.spcm_dwInvalidateBuf(self.hCard, buf_type)
+
+    def _def_transfer64bit(self, buffer_type, direction, bytes_till_event, data_pointer, offset, buffer_length):
+        """Define a 64-bit transer between the device and the computer."""
+        pyspcm.spcm_dwDefTransfer_i64(self.hCard, buffer_type, direction, bytes_till_event, data_pointer, offset, buffer_length)
+
+    def get_max_sample_rate(self, verbose=0):
+        """Return max sample rate."""
+        # read type, function and sn and check for D/A card
+        value = self._param32bit(pyspcm.SPC_PCISAMPLERATE)
+        if verbose:
+            print('max_sample_rate: %s' % (value))

```

Comment:
@eendebakpt  and this, unless there is a good reason to use print() here  that I don't understand :dagger: